ramips: add support for WAVLINK WL-WN577A2 (aka Maginon WLR-755)

lrswss opened the pull request:
This commit adds support for the Wavlink WL-WN577A2 (black case). In Germany this dual-band wall-plug
wireless router is sold under the brand name Maginon WL-755 (white case):

- CPU: MediaTek MT7628AN (580MHz)
- Flash: 8MB
- RAM: 64MB
- Bootloader: U-Boot
- Ethernet: 2x 10/100 Mbps (Ralink RT3050)
- 2.4 GHz: 802.11b/g/n SoC
- 5 GHz: 802.11a/n/ac MT7610E
- Antennas: internal
- 4 green LEDs: 1 programmable (WPS) + LAN, WAN, POWER
- Buttons: Reset, WPS
- Small sliding power switch

Flashing instructions (U-boot):

- Configure a TFTP server on your PC/Laptop and set its IP to 192.168.10.100
- Rename the OpenWrt image as firmware.bin and place it in the root folder of the TFTP server
- Power off the device and connect an Ethernet cable from its LAN or WAN port to your PC/Laptop
- Press the WPS button (and keep it pressed)
- Power on the device (using the small sliding power switch)
- After a few seconds, when the WAN/LAN LED stops blinking very fast, release the WPS button
- Flashing takes about a minute
- WPS LED will indicate OpenWrt device status

lrswss committed:
ramips: add support for WAVLINK WL-WN577A2

adrianschmutzler commented:
Please add the description to the commit message and add a Signed-off-by.

https://openwrt.org/submitting-patches

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,117 @@
+/dts-v1/;
```
Please add an SPDX license identifier before this and an empty line after.

lrswss replied:
Add empty link. But I'm confused, other dts don't have an SPDX license identifier...

adrianschmutzler replied:
Again, we have started to require this add a certain point. And it's not so easy to add (=change) a license on an existing file.

lrswss replied:
Ok, so adding /MIT/ right after /dts-v1/ would fulfill the new requirement? I just grep'ed through the dts files and didn't find one with and license identifier...

lrswss replied:
Or should it state /SPDX-License-Identifier: GPL-2.0-only/?

adrianschmutzler replied:
You will end up with something like this:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7628an_tplink_tl-wr841n-v14.dts

I would call that the preferred license as well. But you may choose your own license identifier (as far as it's reasonable).

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,117 @@
+	model = "WAVLINK WL-WN577A2";
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
```
That's already in mt7628an.dtsi ...

lrswss replied:
Ok, I'll remove it. My dts is based on mt7628an_wavlink_wl-wn575a3.dts...

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,117 @@
+
+&state_default {
+	gpio {
+        groups = "i2c", "wled_an", "gpio", "refclk", "wdt", "p0led_an";
```
I only see i2c wdt p0led_an required?
Despite , indent needs to be fixed.

lrswss replied:
As stated before, the dts is based on mt7628an_wavlink_wl-wn575a3.dts. I tried to find documentation on 'wled_an' and 'p0led_an' but gave up clueless, so I just kept this line unchanged. So you saying that only 'i2c' is required?

lrswss replied:
Removed all but i2c, wdt and p0led_an, reflashed the devices and every seems to be ok. But could you please enlighten about this section of the dts. What is  'p0led_an' and 'wled_an' all about?

adrianschmutzler replied:
you need "i2c", "wdt", "p0led_an"
This depends on the gpio numbers you are assigning in your DTS.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,117 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Please check whether a higher frequency can be used.

lrswss replied:
Other Wavlink similiar devices (WL-WN575, WL-WN570) are also using this frequency so why should I bother changing it? Actually I don't want to brick my device.

adrianschmutzler replied:
That's why most of the old devices just have these 10 MHz, because it was copied over and over ...

Using a higher frequency might speed up reading from flash by a factor 2 to 5.

lrswss replied:
I actually opened the device and the flash chip is labeled AH1943 / 25Q64CSIG3. So, would it be save to try 20MHz as SPI frequency?

lrswss replied:
Ok, increased SPI to 40MHz, reflashed devices, working. :-)

adrianschmutzler commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -537,6 +537,17 @@ define Device/wavlink_wl-wn575a3
+  DEVICE_ALT0_VENDOR := Maginon
+  DEVICE_ALT0_MODEL := WLR-755
+  DEVICE_PACKAGES := kmod-mt76x0e
+  SUPPORTED_DEVICES += wl-wn577a2
```
This line can be dropped.

lrswss replied:
ACK

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -38,6 +38,7 @@ ramips_setup_interfaces()
 			"0:lan" "1:lan" "2:lan" "3:lan" "4:wan" "6@eth0"
 		;;
+	wavlink,wl-wn577a2|\
 	buffalo,wcr-1166ds)
```
Please change the order here, so alphabetic sorting is maintained:
	buffalo,wcr-1166ds|\
	wavlink,wl-wn577a2)

lrswss replied:
ACK

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -184,6 +185,11 @@ ramips_setup_macs()
 		;;
+	wavlink,wl-wn577a2)
+		wan_mac=$(mtd_get_mac_binary factory 0x2e)
+		lan_mac=$(mtd_get_mac_binary factory 0x28)
```
You already have the LAN address set in DTS, so you can drop it here.

lrswss replied:
ACK

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -184,6 +185,11 @@ ramips_setup_macs()
+	wavlink,wl-wn577a2)
+		wan_mac=$(mtd_get_mac_binary factory 0x2e)
+		lan_mac=$(mtd_get_mac_binary factory 0x28)
+		label_mac=$(mtd_get_mac_binary factory 0x4)
```
Can you provide an overview of addresses and increments?
Is the address in 0x4 found in another location as well?

lrswss replied:
Looking at the hexdump of the 'factory' partition I identified the following MAC addresses. 
2.4GHz and 5GHz MACs are actually printed on a label on the device.

0x4   80:3f:5d::xx:xx:xx (2.4GHz)
0x2e   80:3f:5d::xx:xx:xx (LAN)
0x28   80:3f:5d::xx:xx:xx (WAN)
0x8004 80:3f:5d:xx:xx:xx (5GHz)

adrianschmutzler replied:
Well, thanks, but what I need is the last byte (to see the increments).
What I'm actually looking for is whether 0x4 and 0x28 have the same address by coincidence, so we can move label MAC address to DTS.

lrswss replied:
No, 0x4 and 0x28 have different values:
0x8004 80:3f:5d:xx:xx:c2 (5GHz)
0x04   80:3f:5d:xx:x:c1 (2.4GHz)
0x2e   80:3f:5d:xx:xx:c0 (LAN)
0x28   80:3f:5d:xx:xx:bf (WAN)

adrianschmutzler replied:
Fine, then the statement here is correct.

lrswss committed:
ramips: add support for WAVLINK WL-WN577A2 (aka Maginon WLR-755)

This commit adds support for the Wavlink WL-WN577A2 (black case) dual-band
wall-plug wireless router. In Germany this device is sold under the brand
name Maginon WL-755 (white case):

Device specifications:

- CPU: MediaTek MT7628AN (580MHz)
- Flash: 8MB
- RAM: 64MB
- Bootloader: U-Boot
- Ethernet: 2x 10/100 Mbps (Ralink RT3050)
- 2.4 GHz: 802.11b/g/n SoC
- 5 GHz: 802.11a/n/ac MT7610E
- Antennas: internal
- 4 green LEDs: 1 programmable (WPS) + LAN, WAN, POWER
- Buttons: Reset, WPS
- Small sliding power switch

Flashing instructions (U-boot):

- Configure a TFTP server on your PC/Laptop and set its IP to 192.168.10.100
- Rename the OpenWrt image to firmware.bin and place it in the root folder
  of the TFTP server
- Power off (using the small sliding power switch on the left side) the device
  and connect an ethernet cable from its LAN or WAN port to your PC/Laptop
- Press the WPS button (and keep it pressed)
- Power on the device (using the small power switch)
- After a few seconds, when the WAN/LAN LED stops blinking very fast, release
  the WPS button
- Flashing OpenWrt takes less than a minute, system will reboot automatically
- After reboot the WPS LED will indicate the current OpenWrt running status

Signed-off-by: Lars Wessels <software@bytebox.org>

blocktrron commented:
Are Power and LAN/WAN LEDs working as expected?

lrswss commented:
> Are Power and LAN/WAN LEDs working as expected?

Yes, if you plug in an ethernet cable the corresponding LED starts flashing rapidly.

/sys/class/leds:
mt76-phy0 -> ../../devices/platform/10300000.wmac/leds/mt76-phy0/
mt76-phy1 -> ../../devices/pci0000:00/0000:00:00.0/0000:01:00.0/leds/mt76-phy1/

As far as I can tell the assignment is done automatically.
Or is it possible to use these two LEDs for other purposes?

blocktrron commented:
@lrswss The MT7628 has pin-multiplex which can be either configured for specific functions or as GPIO pins. There's an overview with the pin-mapping online in the datasheet (p.23 - Google MT7628 datasheet).

Adding `p3led_an` and `p4led_an` to the `gpio` pingroup and create two LEDs in the device tree should make the LEDs able to be toggled using sysfs (GPIO40 and GPIO39). Can you try that? This consumes some CPU cycles, however it allows people to assign LED functions themselves.

lrswss commented:
> Adding `p3led_an` and `p4led_an` to the `gpio` pingroup and create two LEDs in the device tree should make the LEDs able to be toggled using sysfs (GPIO40 and GPIO39). Can you try that? This consumes some CPU cycles, however it allows people to assign LED functions themselves.

Thank you for sheding some light on this rather complicated matter. I'll try to create an image with the two parameters added and give you some feedback. 

Just to be sure, the corresponding dts file for the wavlink/maginon device should look (in part) like this, right?

```
    leds {
        compatible = "gpio-leds";

        led_wps: wps {
            label = "wl-wn577a2:green:wps";
            gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
        };

        led3 {
            label = "wl-wn577a2:green:led3";
            gpios = <&gpio 40 GPIO_ACTIVE_LOW>;
        };

        led4 {
            label = "wl-wn577a2:green:led4";
            gpios = <&gpio 39 GPIO_ACTIVE_LOW>;
        };
   };
```

```
&state_default {
    gpio {
        groups = "i2c", "wdt", "p0led_an", "p3led_an", "p4led_an";
        function = "gpio";
    };
};
```

What about the two still missing "p1led_an" and "p2led_an" entries? And what is the "p0led_an" used for anyway? I've also seen "wled_an"...

lrswss commented:
Working! :-) Both led3 (LAN) and led4 (WAN) can be assigned using the "LED Configuration" option.

```
mt76-phy0 -> ../../devices/platform/10300000.wmac/leds/mt76-phy0/
mt76-phy1 -> ../../devices/pci0000:00/0000:00:00.0/0000:01:00.0/leds/mt76-phy1/
wl-wn577a2:green:led3 -> ../../devices/platform/leds/leds/wl-wn577a2:green:led3/ 
wl-wn577a2:green:led4 -> ../../devices/platform/leds/leds/wl-wn577a2:green:led4/
wl-wn577a2:green:wps -> ../../devices/platform/leds/leds/wl-wn577a2:green:wps/
```

However mt76-phy0 and mt76-phy1 still show up in the LED drop down menu BUT both cannot be assigned to any function. They just stay off. Is there any (simple) way to remove them from them drop menu?

adrianschmutzler commented:
Congrats. Then please name them accordingly:

```
        lan {
            label = "wl-wn577a2:green:lan";
            gpios = <&gpio 40 GPIO_ACTIVE_LOW>;
        };

        wan {
            label = "wl-wn577a2:green:wan";
            gpios = <&gpio 39 GPIO_ACTIVE_LOW>;
        };
```

You can then set their default usage by mt76x8/base-files/etc/board.d/01_leds

lrswss commented:
> You can then set their default usage by mt76x8/base-files/etc/board.d/01_leds

Already adjusted the DTS accordingly. Decided to assign them to LAN / WAN port to emulate expected behaviour. Just flashed the image and it's working as expected. :-)

```
wavlink,wl-wn577a2)
    ucidef_set_led_switch "lan" "lan" "$boardname:green:lan" "switch0" "0x8"
    ucidef_set_led_switch "wan" "wan" "$boardname:green:wan" "switch0" "0x10"
    ;;
```
How can I remove mt76-phy0 and mt76-phy1 from the LED options menu? What is 'p0led_an' used for? Does is still need to be included in state_default / gpio/ groups?

lrswss committed:
ramips: dts update for WAVLINK WL-WN577A2

- make LAN/WAN leds configurable
- LAN/WAN leds are assigned to switch0 to emulate expected behaviour
- disabled unused usbphy, ehci, ohci in device tree

Signed-off-by: Lars Wessels <software@bytebox.org>

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -36,6 +36,16 @@
 	leds {
 		compatible = "gpio-leds";
 
+		led_lan: lan {
```
you can drop the DT label (led_lan), you don't use it for anything. Same for led_wan.

lrswss replied:
Is that mandatory ? I'd like to stick with the labels since the wps has one as well.

adrianschmutzler replied:
Well, the idea is to only add a label when it is used. The led_wps label is actually used for referencing that led in aliases.

On the other hand, they won't hurt. So, you may as well keep them, but be prepared for somebody else to remove them anyway during a later cleanup.

adrianschmutzler commented:
> What is 'p0led_an' used for?

p0led_an is gpio 43, i.e. the reset button.
wdt is gpio 38, i.e. the wps button.

blocktrron commented:
@lrswss mt76 always registers to the LED subsystem, so oth LEDs are not really an issue.

lrswss commented:
> > What is 'p0led_an' used for?
> p0led_an is gpio 43, i.e. the reset button.
> wdt is gpio 38, i.e. the wps button.

Could you please point me to some documentation on this? I read quite a few articles on Linux device tree and DTS files but never came across 'wled_an', 'pXled', etc. Furthermore parts of the DTS files in the OpenWRT source tree seem to have undergone quite a few changes from 19.0.7 to current master branch (e.g. &gpio1 5 -> &gpio 37). Very confusing...

blocktrron commented:
@lrswss The mt76x8 has three 32bit wide banks for controlling the GPIOs. Previously, there were 2 GPIO controllers in OpenWrt mapped to these banks (Controller 0 - PIN 0-31; Controller 1 PIN 32-63). Now it's a single GPIo controller handling all banks.

Available pingroups on the MT7628 are listed here: https://elixir.bootlin.com/linux/latest/source/arch/mips/ralink/mt7620.c#L307
The mapping GPIO <--> Pingroup is documented in the MT7628 datasheet which can be found via Google.

blocktrron commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,137 @@
+};
+
+&esw {
+	mediatek,portmap = <0x2f>;
```
Tabs vs. spaces

lrswss replied:
ACK

blocktrron commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,137 @@
+};
+
+&ehci {
+    status = "disabled";
```
Tabs vs. spaces

lrswss replied:
ACK

blocktrron commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,137 @@
+};
+
+&usbphy {
+    status = "disabled";
```
Tabs vs. spaces

lrswss replied:
ACK

blocktrron commented on target/linux/ramips/dts/mt7628an_wavlink_wl-wn577a2.dts:
```diff
@@ -0,0 +1,137 @@
+};
+
+&ohci {
+    status = "disabled";
```
Tabs vs. spaces

lrswss replied:
ACK

blocktrron reviewed (approved):
@lrswss I've merged this with minor changes to my staging tree. Can you have a look and approve these?

https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=summary

rotanid commented:
this PR is now integrated in OpenWrt Master branch:
https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=dbaf0d20bb6afec4a592456af6b1238b46fa1acb
